File: 83_remove_duplicates_from_sorted_list/83_remove_duplicates_from_sorted_list.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

##########################################################################
#	> File Name: 83_remove_duplicates_from_sorted_list.py
#	> Author: Tingjian Lau
#	> Mail: tjliu@mail.ustc.edu.cn
#	> Created Time: 2016/05/16
#########################################################################

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        p = head
        
        while p:
            if not p.next:
                break
            if p.val == p.next.val:
                p.next = p.next.next
            else:
                p = p.next
        
        return head

File: 83_remove_duplicates_from_sorted_list/83_remove_duplicates_from_sorted_list.cpp
/*************************************************************************
	> File Name: 83_remove_duplicates_from_sorted_list.cpp
	> Author: Tingjian Lau
	> Mail: tjliu@mail.ustc.edu.cn
	> Created Time: 2016/05/16
 ************************************************************************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode* p = head;
        
        while(p){
            if(!p->next) break; // 在while内进行判断可以省去判断链表为空的情况
            
            if(p->val == p->next->val){ // 这种写法可以避免申请两个节点指针
                p->next = p->next->next;
            }else{
                p = p->next;
            }
        }
        
        return head;
    }
};

